File: frontend/src/api/sample.ts
/**
 * 样品登记（Sample）接口封装 —— T-301
 *
 * 契约见 docs/api/api-spec.md 样品域（/api/sample/*）：
 *   - 导入 POST   /sample/import   （multipart，字段名 file）   权限 sample:import
 *   - 分页 GET    /sample/page                                  权限 sample:query
 *   - 详情 GET    /sample/{id}                                  权限 sample:query
 *   - 维护 PUT    /sample                                       权限 sample:import
 *   - 确认 POST   /sample/confirm  （S10→S20，批量）            权限 sample:confirm
 * 统一响应 { code, msg, data }，分页 data.records/total/current/size，字段 camelCase。
 */
import { get, post, put } from '@/utils/request'
import type { PageResult } from '@/types/api'

/** 样品状态机（AGENTS 7.2，与后端 common/enums/SampleStatus 的 code 严格一致，禁止私改） */
export const SAMPLE_STATUS_OPTIONS = [
  { code: 10, label: '已登记' },
  { code: 20, label: '登记确认' },
  { code: 30, label: '已分解' },
  { code: 40, label: '已安排' },
  { code: 50, label: '检验中' },
  { code: 60, label: '检验完成' },
  { code: 70, label: '已审核' },
  { code: 80, label: '已签发' },
  { code: 90, label: '已出报告' },
] as const

/** 状态 code → 标签色（仅体验层展示） */
export const SAMPLE_STATUS_TAG: Record<number, 'info' | 'primary' | 'success' | 'warning' | 'danger'> = {
  10: 'info',
  20: 'primary',
  30: 'primary',
  40: 'warning',
  50: 'warning',
  60: 'success',
  70: 'success',
  80: 'success',
  90: 'success',
}

/** 样品登记（与后端 entity/VO 对齐，camelCase） */
export interface Sample {
  id?: number
  /** 样品编号（唯一，如 JK(2023)-SA-001） */
  sampleNo: string
  sampleName: string
  /** 受检单位 */
  clientName?: string
  /** 抽样地址 */
  samplingAddress?: string
  payee?: string
  /** 费用 */
  fee?: number
  /** 样品数量（文本，如 3kg） */
  sampleQuantity?: string
  /** 项目名称（如 市级例行） */
  projectName?: string
  /** 采样日期 yyyy-MM-dd */
  samplingDate?: string
  remark?: string
  sampler?: string
  /** 生产单位 */
  manufacturer?: string
  /** 抽样基数 */
  samplingBase?: string
  /** 样品状态（如 鲜活） */
  sampleState?: string
  spec?: string
  brand?: string
  grade?: string
  /** 原编号或生产日期 */
  originalNo?: string
  /** 检验类别（如 监督抽检） */
  inspectType?: string
  /** 要求完成日期 yyyy-MM-dd */
  requireCompleteDate?: string
  /** 关联监抽任务编号 */
  taskNo?: string
  taskBatchNo?: string
  /** 样品状态机 code（S10=10 … S90=90） */
  status: number
  /** 状态中文名（后端只读输出） */
  statusLabel?: string
  /** 登记确认人 / 时间（只读） */
  confirmedBy?: string
  confirmedAt?: string
  /** 审计字段（只读） */
  createdBy?: string
  createdAt?: string
  updatedAt?: string
}

/** 列表查询参数 */
export interface SampleQuery {
  pageNum: number
  pageSize: number
  sampleNo?: string
  sampleName?: string
  taskNo?: string
  status?: number
}

/** 导入失败明细行 */
export interface SampleImportErrorRow {
  /** Excel 行号（1 基） */
  rowNum: number
  /** 样品编号（可空） */
  sampleNo?: string
  message: string
}

/** 导入结果 */
export interface SampleImportResult {
  total: number
  successCount: number
  failCount: number
  errors: SampleImportErrorRow[]
}

/** 分页查询样品：GET /sample/page */
export function pageSampleApi(params: SampleQuery): Promise<PageResult<Sample>> {
  return get<PageResult<Sample>>('/sample/page', params as unknown as Record<string, unknown>)
}

/** 样品详情：GET /sample/{id} */
export function getSampleApi(id: number): Promise<Sample> {
  return get<Sample>(`/sample/${id}`)
}

/** 采样单 Excel 导入：POST /sample/import（multipart，字段名 file） */
export function importSampleApi(file: File): Promise<SampleImportResult> {
  const formData = new FormData()
  formData.append('file', file)
  return post<SampleImportResult>('/sample/import', formData)
}

/** 登记信息维护：PUT /sample */
export function updateSampleApi(payload: Sample): Promise<void> {
  return put<void>('/sample', payload)
}

/** 登记确认（S10→S20，批量）：POST /sample/confirm */
export function confirmSampleApi(ids: number[]): Promise<{ confirmedCount: number }> {
  return post<{ confirmedCount: number }>('/sample/confirm', { ids })
}

File: backend/src/main/java/com/lims/entity/Sample.java
package com.lims.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.lims.common.enums.ResultConclusion;
import com.lims.common.enums.SampleStatus;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * 样品登记（sample_info，T-301）。
 *
 * <p>字段与采样单 Excel 列头一一对应（db/init/05_sample_tables.sql）。</p>
 *
 * <p>表名用 {@code sample_info} 而非 {@code sample}：`SAMPLE` 是 SQL 关键字，
 * 与 MyBatis-Plus 分页插件的 JSqlParser 冲突，会令 count SQL 优化失败
 * （DECISIONS 2026-09-11；同 sys_user 避开 user 关键字的处理原则）。</p>
 *
 * <p>{@code status} 为 {@link SampleStatus} 枚举，落库 TINYINT code、JSON 输出数字 code
 * （AGENTS 4.3「业务状态字段用枚举，禁止魔法数字」）；中文名另经 {@link #getStatusLabel()} 出网。
 * 状态变更必须经 SampleStatusTransition 白名单校验，DAO 层禁止直接 set。</p>
 */
@Data
@EqualsAndHashCode(callSuper = true)
@TableName("sample_info")
public class Sample extends BaseEntity {

    @TableId(type = IdType.AUTO)
    private Long id;

    /** 样品编号（唯一，如 JK(2023)-SA-001） */
    private String sampleNo;

    private String sampleName;

    /** 受检单位 */
    private String clientName;

    /** 抽样地址 */
    private String samplingAddress;

    private String payee;

    private BigDecimal fee;

    /** 样品数量（文本，如 3kg） */
    private String sampleQuantity;

    /** 项目名称（如 市级例行） */
    private String projectName;

    /** 采样日期（Excel「日期」列） */
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate samplingDate;

    private String remark;

    private String sampler;

    /** 生产单位 */
    private String manufacturer;

    /** 抽样基数 */
    private String samplingBase;

    /** 样品状态（如 鲜活） */
    private String sampleState;

    /** 规格型号 */
    private String spec;

    private String brand;

    /** 样品等级 */
    private String grade;

    /** 原编号或生产日期 */
    private String originalNo;

    /** 检验类别（如 监督抽检） */
    private String inspectType;

    /** 要求完成日期 */
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate requireCompleteDate;

    /** 关联监抽任务编号（supervise_task.task_no） */
    private String taskNo;

    /** 任务批号 */
    private String taskBatchNo;

    /** 样品状态机（落库 TINYINT code；导入落库即 S10） */
    private SampleStatus status;

    /**
     * 整体结论（T-601）：1=合格 2=不合格 3=待判定。
     *
     * <p>AGENTS 7.3 规则 6 的派生值——由「该样品全部**非参考项**单项结论」聚合而来
     * （参考项不计入整体，白名单 D3 裁决）。在结果保存/提交/查询时重算回写，
     * 供报告生成（T-702）与查询（T-801）直接取用。未进入录入阶段时为 {@code null}。</p>
     */
    private ResultConclusion conclusion;

    /** 登记确认人（S10→S20 时写入） */
    private String confirmedBy;

    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime confirmedAt;

    // ---------------------------------------------------------------------
    // 审核 / 签发（T-701：S60→S70→S80）
    // 完整流水见 sample_audit_log；本处保存**当前有效值**，供 T-702 报告合成直接取用
    // （业务说明书报告页脚：「报告无制表、审核、批准人签字无效」）。
    // ---------------------------------------------------------------------

    /** 审核人工号（S60→S70 写入；审核退回时清空） */
    private String auditBy;

    /** 审核时间 */
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime auditAt;

    /** 审核意见（可空；审核退回时原因记入 sample_audit_log.opinion） */
    private String auditOpinion;

    /** 签发人工号（S70→S80 写入） */
    private String signBy;

    /** 签发时间 */
    @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")
    private LocalDateTime signAt;

    /** 状态中文名（非持久化，出网供前端展示，避免前端维护 code→label 字典） */
    public String getStatusLabel() {
        return status == null ? null : status.getLabel();
    }

    /** 整体结论中文名（非持久化，出网供前端展示） */
    public String getConclusionLabel() {
        return conclusion == null ? null : conclusion.getLabel();
    }
}
